src/parser.cpp

```diff
@@ -0,0 +1,637 @@
+{
+    rhs.mPimpl = nullptr;
+}
+
+Parser& Parser::operator=(Parser p)
+{
+    Logger::operator =(p);
+    p.swap(*this);
+    return *this;
+}
+
+void Parser::swap(Parser &rhs)
+{
+    std::swap(this->mPimpl, rhs.mPimpl);
+}
+
+ModelPtr Parser::parseModel(const std::string &input)
+{
+    ModelPtr model = std::make_shared<Model>();
+    updateModel(model, input);
+    return model;
+}
+
+void Parser::updateModel(const ModelPtr &model, const std::string &input)
+{
+    if (mPimpl->mFormat == Format::XML) {
+        XmlDocPtr doc = std::make_shared<XmlDoc>();
+        doc->parse(input);
+        // Copy any XML parsing errors into the common error handler.
+        if (doc->xmlErrorCount() > 0) {
+            for (size_t i = 0; i < doc->xmlErrorCount(); ++i) {
+                EntityErrorPtr err = std::make_shared<EntityError>();
+                err->setDescription(doc->getXmlError(i));
+                addError(err);
+            }
+        }
+        const XmlNodePtr node = doc->getRootNode();
+        if (node->isType("model")) {
+            loadModel(model, node);
+        } else {
+            ModelErrorPtr err = std::make_shared<ModelError>();
+            err->setDescription("Unexpected XML element type (not a model).");
+            err->setModel(model);
+            addError(err);
+        }
+    }
+}
+
+void Parser::loadModel(const ModelPtr &model, const XmlNodePtr &node)
+{
+    XmlAttributePtr attribute = node->getRootAttribute();
+    while (attribute) {
+        if (attribute->isType("name")) {
+            model->setName(attribute->getValue());
+        } else {
+            ModelErrorPtr err = std::make_shared<ModelError>();
+            err->setDescription("Invalid model attribute: '" + attribute->getType() +
+                                "' with value: '" + attribute->getValue() + "'.");
+            err->setModel(model);
+            addError(err);
+        }
+        attribute = attribute->getNext();
+    }
+    XmlNodePtr childNode = node->getChild();
+    while (childNode) {
+        if (childNode->isType("component")) {
+            ComponentPtr component = std::make_shared<Component>();
+            loadComponent(component, childNode);
+            model->addComponent(component);
+        } else if (childNode->isType("units")) {
+            UnitsPtr units = std::make_shared<Units>();
+            loadUnits(units, childNode);
+            model->addUnits(units);
+        } else if (childNode->isType("import")) {
+            ImportPtr import = std::make_shared<Import>();
+            loadImport(import, model, childNode);
+        } else if (childNode->isType("encapsulation")) {
+            XmlNodePtr componentRefNode = childNode->getChild();
+            if (componentRefNode) {
+                loadEncapsulation(model, componentRefNode);
+            } else {
+                ModelErrorPtr err = std::make_shared<ModelError>();
+                err->setDescription("Encapsulation does not contain any component_ref");
```

Comment:
Missing full stop for the description. Otherwise, I couldn't find a test for this case and although the coverage test does indeed cover this case, I would expect a test for this case. (Maybe the same with other similar bits of code?)